Convert Tendermint Block Response

RiccardoM opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Converts a `ResultBlock` into a `CosmosResultBlock` having validator addresses represented as Bech32 addresses instead of Hex addresses.

Closes #3544

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

RiccardoM commented:
Should I write some tests for this? If so, which ones?

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+type CosmosBlock struct {
+	mtx          sync.Mutex
```
woaaa, why is there a mutex?

alexanderbez commented on go.mod:
```diff
@@ -6,11 +6,11 @@ require (
 	github.com/btcsuite/btcd v0.0.0-20190115013929-ed77733ec07d
 	github.com/cosmos/go-bip39 v0.0.0-20180618194314-52158e4697b8
 	github.com/cosmos/ledger-cosmos-go v0.10.3
+	github.com/cosmos/tools/cmd/runsim v0.0.0-20190816161256-f506590651e9 // indirect
```
undo

RiccardoM replied:
Done

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,147 @@
+	"time"
+)
+
+type CosmosResultBlock struct {
```
Can we think of better type prefixes? Maybe even just remove `Cosmos*`.

marbar3778 replied:
https://github.com/tendermint/tendermint/blob/master/rpc/core/types/responses.go#L28

RiccardoM replied:
Done

tnachen commented on client/rpc/block.go:
```diff
@@ -2,12 +2,11 @@ package rpc
 	"github.com/gorilla/mux"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"net/http"
```
Please move this back.

RiccardoM replied:
What do you mean by that?

tnachen replied:
Actually you just need to run goimport to reformat this.

fedekunze replied:
i.e run `make format`

RiccardoM replied:
Sorry, was using gofmt instead of goimports. Fixed it now

marbar3778 commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,147 @@
+
+type CosmosBlockMeta struct {
+	BlockID types.BlockID `json:"block_id"` // the block hash and partsethash
+	Header  CosmosHeader  `json:"header"`   // The block's Header
```
```suggestion
	Header  types.Header  `json:"header"`   // The block's Header
```

this may better to not have to update

marbar3778 commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,147 @@
+	Header  CosmosHeader  `json:"header"`   // The block's Header
+}
+
+type CosmosHeader struct {
```
if Tendermint header updates I believe it will break this?

RiccardoM replied:
Probably, but IMO it's the only way we can have Bech32 validator addresses into the block header. 

Do you have any suggestion on how we can achieve that otherwise without replacing the header type definition?

RiccardoM replied:
I think I got it and it's being discussed inside https://github.com/cosmos/cosmos-sdk/pull/4922#discussion_r315697469. 

Can we close this?

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,146 @@
+	"github.com/tendermint/tendermint/version"
+)
+
+type ResultBlock struct {
```
Fun fact, we don't need to reimplement/redefine everything! Instead, define the wrapper types and embed the original struct. Then, include any fields you wish to override. But because amino doesn't comply with the JSON stdlib, you'll have to implement a custom marshaler.

e.g.

```go
// Header defines a wrapper around Tendermint's Header type overriding various
// fields.
type Header struct {
	// embed original type
	tmtypes.Header

	// override fields from original type
	ProposerAddress sdk.ValAddress `json:"proposer_address"`
}

// MarshalJSON implements the json.Marshaler interface. We do this because Amino
// does not respect the JSON stdlib embedding semantics.
func (h Header) MarshalJSON() ([]byte, error) {
	type headerJSON Header
	_h := headerJSON(h)

	return json.Marshal(_h)
}
```

RiccardoM replied:
Thanks for the tip, I've implemented as you suggested. I also used the same method for `CommitSig`. Could you please check if that's ok and/or could be used elsewhere?

RiccardoM commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,113 @@
+	types.Header
+
+	// override fields from original type
+	ProposerAddress sdk.ValAddress `json:"proposer_address"`
```
@alexanderbez can we ignore these issues?

alexanderbez replied:
Ahh yes, we need to add a `// nolint: structtag` above the types to ignore this.

RiccardoM replied:
Done

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,113 @@
+
+type CommitSig struct {
+	types.CommitSig
+	ValidatorAddress sdk.ValAddress `json:"validator_address"`
```
ditto

RiccardoM replied:
Done

RiccardoM commented on client/rpc/block.go:
```diff
@@ -68,7 +68,10 @@ func getBlock(cliCtx context.CLIContext, height *int64) ([]byte, error) {
+	// convert to new type
+	converted := ConvertBlockResult(res)
+
+	return codec.Cdc.MarshalJSON(converted)
```
Done

RiccardoM commented on go.sum:
```diff
@@ -7,6 +7,8 @@ github.com/aead/siphash v1.0.1/go.mod h1:Nywa3cDsYNNK3gaciGTWPwHt0wlpNV15vwmswBA
 github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
 github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf/go.mod h1:ybxpYRFXyAe+OPACYpWeL0wqObRcbAqCMya13uyzqw0=
 github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6/go.mod h1:grANhF5doyWs3UAsr3K4I6qtAmlQcZDesFNEHPZAzj8=
+github.com/aws/aws-sdk-go v1.20.20 h1:OAR/GtjMOhenkp1NNKr1N1FgIP3mQXHeGbRhvVIAQp0=
```
Done

fedekunze commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,115 @@
+	"github.com/tendermint/tendermint/types"
+)
+
+type ResultBlock struct {
```
can we add godoc comments on each of the public structs and function?

RiccardoM replied:
Added. Also unified how different types are written. Could you check please?

RiccardoM replied:
Added

fedekunze commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,115 @@
+
+func ConvertBlockResult(res *ctypes.ResultBlock) (blockResult ResultBlock) {
+
+	// header
```
no need for all these comments

RiccardoM replied:
Removed

marbar3778 commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,115 @@
+	Block     Block     `json:"block"`
+}
+
+type BlockMeta struct {
```
nitpick, to follow the same structure as header. `types.BlockMeta`, not blocking the merge though.

RiccardoM replied:
Unified all the types. Should be good now

fedekunze commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,110 @@
+	"github.com/tendermint/tendermint/types"
+)
+
+// Single block (with meta)
```
```suggestion
// ResultBlock represents a single block with metadata
```

fedekunze commented:
bump @RiccardoM

RiccardoM commented:
I've added a simple test. However, I would like to have a test that you can make sure works properly. For this I would need 

- A default block resposne
- The mapping between the hash address and its Bech32 representation

I'm also not able to reproduce what @alexanderbez said about the response not having the data inside it

alexanderbez commented:
@RiccardoM I recommend you launch a local network and compare block responses pre and post changes. You'll see it's missing a bunch of the fields.

RiccardoM commented:
@alexanderbez thanks for the tip. I've updated the code to fix those errors. 
I've also added the overriding of `int`, `int64` and `uint64` fields in order to preserve their serialization as strings.

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,223 @@
+
+	// override fields from original type
+	Version         Consensus      `json:"version"`
+	Height          int64          `json:"height,string"`    // Override int64 fields do to json.Marshal not converting them to string
```
Instead of copying the side comment, I would just add a single comment to the top and group them:


```go
type Header struct {
  // override fields so json.Marshal will marshal in accordance with amino JSON format
  Version         Consensus
  Height          int64          `json:"height,string"`
  // ...

  ProposerAddress sdk.ValAddress `json:"proposer_address"`
}

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,223 @@
+func (h Header) MarshalJSON() ([]byte, error) {
+	type headerJSON Header
+	_h := headerJSON(h)
+	return json.Marshal(_h)
```
```suggestion
	return json.Marshal(headerJSON(_h))
```

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,223 @@
+// does not respect the JSON stdlib embedding semantics.
+func (h Header) MarshalJSON() ([]byte, error) {
+	type headerJSON Header
+	_h := headerJSON(h)
```
```suggestion
```

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,223 @@
+// does not respect the JSON stdlib embedding semantics.
+func (b BlockID) MarshalJSON() ([]byte, error) {
+	type blockIDJson BlockID
+	_h := blockIDJson(b)
```
same changes as above.

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,223 @@
+// does not respect the JSON stdlib embedding semantics.
+func (b PartSetHeader) MarshalJSON() ([]byte, error) {
+	type partSetHeadJson PartSetHeader
+	_h := partSetHeadJson(b)
```
ditto

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,223 @@
+
+// Consensus defines a wrapper around Tendermint's Consensus type overriding various fields.
+// nolint: structtag
+type Consensus struct {
```
ditto

RiccardoM commented:
@alexanderbez Solved everything. It feels a bit like a shame, but I think it improves a lot the developer UX

RiccardoM commented:
Ping @alexanderbez the issues should be solved. Can we merge this?

RiccardoM commented:
Up @alexanderbez, @jackzampolin @fedekunze. Can we merge this?

alessio commented on client/rpc/block_converter_test.go:
```diff
@@ -0,0 +1,24 @@
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+)
+
+var defaultBlockJson = `{"block_meta":{"block_id":{"hash":"30CD3A9EF2082FF9F2575655E99E37ACC3936DC9E534ADF0BD7436C76258225C","parts":{"total":"1","hash":"463460DA73FA0AE441B6041BF2683AE625CE2D9FD290F4C86CD83D1A7FB9F439"}},"header":{"version":{"block":"10","app":"0"},"chain_id":"test-chain-RoCfX4","height":"16","time":"2019-10-16T09:47:21.054275229Z","num_txs":"0","total_txs":"0","last_block_id":{ ...
```
Please avoid using global variables when writing test cases. This is currently used by one test case only

RiccardoM replied:
Fixed

alessio commented on client/rpc/block_converter_test.go:
```diff
@@ -0,0 +1,24 @@
+	assert.Equal(t, "cosmosvaloper1az62lz2mxqwr6ugge2f6pwfrfcaz5jephqhvkt", convertedBlock.BlockMeta.Header.ProposerAddress.String())
+
+	convertedBlockString := string(cdc.MustMarshalJSON(&convertedBlock))
+	assert.NotContains(t, "E8B4AF895B301C3D7108CA93A0B9234E3A2A4B21", convertedBlockString)
```
I'd like to see a test covering the case with *ctypes.ResultBlock argument being null.

RiccardoM replied:
Added, thanks for the hint

fedekunze commented:
@alessio @alexanderbez mind reviewing again?

alexanderbez commented on client/rpc/block_converter_test.go:
```diff
@@ -0,0 +1,27 @@
+
+	// test when the block is not nil
+	var block ctypes.ResultBlock
+	defaultBlockJSON := `{"block_meta":{"block_id":{"hash":"30CD3A9EF2082FF9F2575655E99E37ACC3936DC9E534ADF0BD7436C76258225C","parts":{"total":"1","hash":"463460DA73FA0AE441B6041BF2683AE625CE2D9FD290F4C86CD83D1A7FB9F439"}},"header":{"version":{"block":"10","app":"0"},"chain_id":"test-chain-RoCfX4","height":"16","time":"2019-10-16T09:47:21.054275229Z","num_txs":"0","total_txs":"0","last_block_id":{"h ...
```
Can we use block JSON that has _all_ fields populated (i.e. txs and evidence). I submitted a tx (using this branch) and when querying the block, I don't see the tx under `block.txs`. I bet the same will hold true for `block.evidence`.

So ideally we have:

```go
// ...

convertedBlock := ConvertBlockResult(&block)
expectedOut := `{...}`

bz, err := cdc.MarshalJSON(convertedBlock)
require.NoError(t, err)
require.Equal(t, expectedOut, string(bz))
```

alexanderbez commented on client/rpc/block_converter.go:
```diff
@@ -0,0 +1,220 @@
+// does not respect the JSON stdlib embedding semantics.
+func (b Block) MarshalJSON() ([]byte, error) {
+	type blockJSON Block
+	return json.Marshal(blockJSON(b))
```
We should probably return a reference to avoid mutex copying.
